### PR TITLE
Use hypothesis

### PR DESCRIPTION
**Pull request by RAMitchell**
Use property based testing with hypothesis library. This replaces tests for CPU and GPU updaters.

The result is reduced code and significantly increased test coverage. This testing pattern can be further extended to other tests over time.

Tests relating to subsampling or column sampling are currently flaky. This shows a bug in xgboost and should be addressed later.

### CODE REVIEW

File: src/data/data.cu [hunk 1]
```diff
@@ -14,7 +14,7 @@
 
 namespace xgboost {
 
-void CopyInfoImpl(ArrayInterface column, HostDeviceVector<float>* out) {
+void CopyFloatImpl(ArrayInterface column, HostDeviceVector<float>* out) {
```
Comment (trivialfis) [inline src/data/data.cu:17]: I think the input array can be any numeric type?

**Comment by RAMitchell** [comment]
@hcho3 how would I add hypothesis to the windows test environment?

**Comment by hcho3** [comment]
@RAMitchell Modify this line
https://github.com/dmlc/xgboost/blob/306e38ff3145b559e3a7c67d071d8a4ced5f70a3/Jenkinsfile-win64#L116
and this line
https://github.com/dmlc/xgboost/blob/306e38ff3145b559e3a7c67d071d8a4ced5f70a3/Jenkinsfile-win64#L141

Right now, we install NumPy, Scipy, and scikit-learn. We should be able to add Hypothesis to the list.

**Comment by hcho3** [comment]
Hmm let’s use `conda install` for the new packages. Also, do we actually need cupy?

**Comment by RAMitchell** [comment]
cupy would allow testing device memory interfaces on Windows. Seems quite useful, but we could skip it if there are issues.

**Comment by hcho3** [comment]
cuPy appears to crash mysteriously.

**Comment by RAMitchell** [comment]
Seems like an issue with in place predict, I'll take a look tomorrow.

**Comment by hcho3** [comment]
I'm not sure how easy it is to support device memory interfaces on Windows. If it's too much hassle we can leave it out for this PR. For the record, RAPIDS library currently only supports Linux.

File: src/data/data.cu [hunk 2]
```diff
@@ -34,6 +34,30 @@ void CopyInfoImpl(ArrayInterface column, HostDeviceVector<float>* out) {
   });
 }
 
+void CopyGroupInfoImpl(ArrayInterface column, std::vector<bst_group_t>* out) {
```
Comment (trivialfis) [inline src/data/data.cu:37]: Let me change group ptr into host device vector later.

File: tests/python-gpu/test_gpu_updaters.py [hunk 3]
```diff
@@ -94,13 +91,11 @@ def test_with_empty_dmatrix(self):
...
-                          'max_bin': [2, 64],
-                          'grow_policy': ['lossguide'],
-                          'tree_method': ['gpu_hist']}
-        for param in parameter_combinations(variable_param):
-            gpu_results = run_suite(param, select_datasets=datasets)
-            assert_results_non_increasing(gpu_results, 1e-2)
+    @given(tm.dataset_strategy)
+    @settings(deadline=None, max_examples=10)
+    def test_specified_gpu_id_gpu_update(self, dataset):
+        param = {'tree_method': 'gpu_hist', 'gpu_id': 1}
```
Comment (trivialfis) [inline tests/python-gpu/test_gpu_updaters.py:97]: Any way that we can make `gpu_id` part of the strategy?

File: tests/python/test_linear.py [hunk 4]
```diff
@@ -1,87 +1,49 @@
...
+
+
+def train_result(param, dmat, num_rounds):
+    result = {}
+    xgb.train(param, dmat, num_rounds, [(dmat, 'train')], verbose_eval=False,
+              evals_result=result)
+    return result
+
+
+# Loss is not guaranteed to always decrease because of regularisation parameters
```
Comment (trivialfis) [inline tests/python/test_linear.py:28]: Can we have a test that when regularization is not applied, the loss is guaranteed to decrease?

Comment (RAMitchell) [reply tests/python/test_linear.py]: Done

File: tests/python-gpu/test_gpu_updaters.py (hunk 3 above)
Comment (RAMitchell) [reply tests/python-gpu/test_gpu_updaters.py]: Done
